### PR TITLE
[New feature] Support hit@k for classification tasks

### PR DESCRIPTION
**Pull request by classicsong**
*Issue #, if available:*
Support hit@k for binary classification tasks. With hit@k, people can check with K most confident predictions, how many are True positives.

*Description of changes:*
In implementation, we sort the prediction values (probabilities of the prediction being 1 before calling argmax) and get the labels of the K predictions with the highest probabilities. Then we count how many samples have labels of 1.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: python/graphstorm/eval/eval_func.py [hunk 1]
```diff
@@ -195,6 +210,38 @@ def labels_to_one_hot(labels, total_labels):
...
+        "Computing hit@K for classification only works for binary classification tasks." \
+        "The preds must be a 2D tensor with the second dimension of 2. " \
+        "The labels must be a 1D tensor or a 2D tensor with the second dimension of 1"
+
+    # preds is a 2D tensor storing
+    # [probability of label 0, probability of label 1]
+    # 0 means negative, 1 means positive.
+    # We compute hit@K for positive labels
+    preds = preds[:,1]
+    if len(labels.shape) == 2:
```
Comment (zhjwy9343) [inline python/graphstorm/eval/eval_func.py:237]: This line will not happen unless remove the `len(labels.shape) == 1` in line #227.

File: tests/unit-tests/test_eval_func.py [hunk 2]
```diff
@@ -377,7 +384,95 @@ def test_compute_per_class_roc_auc():
...
+    try:
+        metric.assert_supported_metric("hit_at_ten")
+        pass_assert = True
+    except:
+        pass_assert = False
+    assert not pass_assert
+
+def test_compute_hit_at_classification():
+    preds = th.arange(100) / 102
+    # preds = [probe_of_0, probe_of_1]
```
Comment (zhjwy9343) [inline tests/unit-tests/test_eval_func.py:446]: Can remove this line?

**Comment by zhjwy9343** [comment]
Can we modify the configuration file `https://graphstorm.readthedocs.io/en/latest/configuration/configuration-run.html#general-configurations` to add this `hit_at_k` metric and its examples?

File: tests/unit-tests/test_eval_func.py (hunk 2 above)
Comment (classicsong) [reply tests/unit-tests/test_eval_func.py:446]: We should keep this to tell how preds looks like.

**Comment by classicsong** [comment]
> Can we modify the configuration file `https://graphstorm.readthedocs.io/en/latest/configuration/configuration-run.html#general-configurations` to add this `hit_at_k` metric and its examples?

Good call out.

File: python/graphstorm/eval/eval_func.py [hunk 3]
```diff
@@ -195,6 +210,38 @@ def labels_to_one_hot(labels, total_labels):
...
+            A 1-D tensor for single-label classification.
+        k: int
+            Hit@K
+    """
+    assert len(preds.shape) == 2 \
+        and preds.shape[1] == 2 \
+        and len(labels.shape) == 1, \
+        "Computing hit@K for classification only works for binary classification tasks." \
+        "The preds must be a 2D tensor with the second dimension of 2. " \
+        "The labels must be a 1D tensor or a 2D tensor with the second dimension of 1"
```
[the remark this answers is not in the document]
Comment (classicsong) [reply python/graphstorm/eval/eval_func.py:230]: Updated

File: docs/source/configuration/configuration-run.rst [hunk 4]
```diff
@@ -325,12 +325,13 @@ General Configurations
...
+- **eval_metric**: Evaluation metric used during evaluation. The input can be a string specifying the evaluation metric to report or a list of strings specifying a list of evaluation metrics to report. The first evaluation metric is treated as the major metric and is used to choose the best trained model. The supported evaluation metrics of classification tasks include ``accuracy``, ``precision_recall``, ``roc_auc``, ``f1_score``, ``per_class_f1_score``, ``hit_at_k``. To be noted, ``hit_at_k`` only works with binary classification tasks. The ``k`` of ``hit_at_k`` can be any positive integer, for example ``hit_at_10`` or ``hit_at_100``. The term ``hit_at_k`` refers to the number of true positives among the top ``k`` predictions with the highest confidence scores. The supported evaluation metrics of regression tasks include ``rmse``, ``mse`` and ``mae``. The supported evaluation metrics of link prediction tasks include ``mrr``.
 
     - Yaml: ``eval_metric:``
         | ``- accuracy``
         | ``- precision_recall``
-    - Argument: ``--eval-metric accuracy precision_recall``
+        | ``- top_at_10``
```
Comment (zhjwy9343) [inline docs/source/configuration/configuration-run.rst:333]: `top_at_10` -> `hit_at_10`?

File: docs/source/configuration/configuration-run.rst [hunk 5]
```diff
@@ -325,12 +325,13 @@ General Configurations
...
+    - Argument: ``--eval-metric accuracy precision_recall top_at_10``
```
Comment (zhjwy9343) [inline docs/source/configuration/configuration-run.rst:334]: ditto

**Review by zhjwy9343 — APPROVED** [review]
LGTM

File: docs/source/configuration/configuration-run.rst (hunk 5 above)
Comment (classicsong) [reply docs/source/configuration/configuration-run.rst]: Updated

**Pull request merged**